packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,121 @@
+import { ModuleManager } from "safe-contracts/base/ModuleManager.sol";
+import { SignatureDecoder } from "safe-contracts/common/SignatureDecoder.sol";
+import { Enum } from "safe-contracts/common/Enum.sol";
+
+contract LivenessGuard is SignatureDecoder, BaseGuard {
+    /// @notice Emitted when a new set of signers is recorded.
+    /// @param signers An arrary of signer addresses.
+    event SignersRecorded(bytes32 indexed txHash, address[] signers);
+
+    Safe public immutable safe;
+    mapping(address => uint256) public lastSigned;
+
+    constructor(Safe _safe) {
+        safe = _safe;
+    }
+
+    /// @notice We just need to satisfy the BaseGuard interfae, but we don't actually need to use this method.
+    function checkAfterExecution(bytes32, bool) external pure {
+        return;
+    }
+
+    /// @notice Records the most recent time which any owner has signed a transaction.
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures,
+        address
+    )
+        external
+    {
+        require(msg.sender == address(safe), "LivenessGuard: only Safe can call this function");
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash(
+            // Transaction info
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            // Payment info
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver,
+            // Signature info
+            Safe(payable(msg.sender)).nonce() - 1
+        );
+        address[] memory signers = _getNSigners(txHash, signatures);
+        for (uint256 i = 0; i < signers.length; i++) {
+            lastSigned[signers[i]] = block.timestamp;
+        }
+        emit SignersRecorded(txHash, signers);
+    }
+
+    /// @notice Exctract the signers from a set of signatures.
+    function _getNSigners(bytes32 dataHash, bytes memory signatures) internal pure returns (address[] memory _owners) {
+        uint256 numSignatures = signatures.length / 65;
+        _owners = new address[](numSignatures);
+
+        /// The following code is extracted from the Safe.checkNSignatures() method. It removes the signature
+        /// validation code, and keeps only the parsing code necessary to extract the owner addresses from the
+        /// signatures. We do not double check if the owner derived from a signature is valid. As this is handled
+        /// in the final require statement of Safe.checkNSignatures().
+        address currentOwner;
```

Comment:
If we could break this into a library and then differential test against the main impl where the code was taken from, that would be amazing

Reply:
https://github.com/ethereum-optimism/optimism/blob/jm/safe-liveness/packages/contracts-bedrock/test/LivenessGuard.t.sol#L39